src/vsr/replica.zig

```diff
@@ -2642,6 +3054,65 @@ pub fn Replica(
             return false;
         }
 
+        /// Returns whether the highest known op is certain.
+        ///
+        /// After recovering the WAL, there are 2 possible outcomes:
+        /// * All entries valid. The highest op is certain, and safe to set as `replica.op`.
+        /// * One or more entries are faulty. The highest op isn't certain — it may be one of the
+        ///   broken entries.
+        ///
+        /// The replica must refrain from repairing any faulty slots until the highest op is known.
+        /// Otherwise, if we were to repair a slot while uncertain of `replica.op`:
+        ///
+        /// * we may nack an op that we shouldn't, or
+        /// * we may replace a prepared op that we were guaranteeing for the primary, potentially
+        ///   forking the log.
+        ///
+        ///
+        /// Test for a fault the right of the current op. The fault might be our true op, and
+        /// sharing our current `replica.op` might cause the cluster's op to likewise regress.
+        ///
+        /// Note that for our purposes here, we only care about entries that were faulty during
+        /// WAL recovery, not ones that were found to be faulty after the fact (e.g. due to
+        /// `request_prepare`).
+        ///
+        /// Cases (`✓`: `replica.op_checkpoint`, `✗`: faulty, `o`: `replica.op`):
+        /// * ` ✓ o ✗ `: View change is unsafe.
+        /// * ` ✗ ✓ o `: View change is unsafe.
+        /// * ` ✓ ✗ o `: View change is safe.
+        /// * ✓=o:       View change is unsafe if any slots are faulty
+        ///              (`replica.op_checkpoint` == `replica.op`).
+        // TODO Use this function once we switch from recovery protocol to the superblock.
+        // If there is an "unsafe" fault, we will need to request a start_view from the leader to
+        // learn the op.
+        fn op_certain(self: *const Self) bool {
+            assert(self.status == .recovering);
+            assert(self.journal.recovered);
+            assert(self.op_checkpoint <= self.op);
+
+            const slot_op_checkpoint = self.journal.slot_for_op(self.op_checkpoint);
+            const slot_op = self.journal.slot_with_op(self.op).?;
+            const slot_op_min = std.math.min(slot_op, slot_op_checkpoint);
+            const slot_op_max = std.math.max(slot_op, slot_op_checkpoint);
+
+            var iterator = self.journal.faulty.bits.iterator(.{ .kind = .set });
+            while (iterator.next()) |slot| {
+                // The command is `reserved` when the entry was found faulty during WAL recovery.
+                // Faults found after WAL recovery are not relevant, because we know their op.
+                if (self.journal.headers[slot.index].command != .reserved) continue;
```

Comment:
There are two mutually exclusive categories of faulty slots here:

1. The slot was found to be faulty _during_ WAL recovery. (For example, there was a power failure, and when the replica restarted it found corruption in the WAL).
2. The slot was found to be faulty _after_ WAL recovery. (For example, another replica requested a prepare. When we read the WAL to retrieve it, we find that it is corrupt).

For the purposes of this method ("knowing" the op) we only care about the former. We differentiate between the two by:

1. In case 1, `journal.headers[slot].command == .reserved`.
2. In case 2, `journal.headers[slot].command == .prepare`.